reasoner/Reasoner.java

```diff
@@ -80,51 +77,49 @@ public Reasoner(ConceptManager conceptMgr, LogicManager logicMgr,
     }
 
     private boolean isInfer(Context.Query context) {
-        return context.options().infer() && !context.transactionType().isWrite();
+        return context.options().infer() && !context.transactionType().isWrite() && logicMgr.rules().hasNext();
     }
 
     public ResourceIterator<ConceptMap> execute(Disjunction disjunction, List<Identifier.Variable.Name> filter,
                                                 Context.Query context) {
         ResourceIterator<Conjunction> conjs = iterate(disjunction.conjunctions());
         if (!context.options().parallel()) return conjs.flatMap(conj -> iterator(conj, filter, context));
-        else return produce(conjs.flatMap(conj -> producers(conj, filter, context)).toList(), context.producer());
+        else return produce(conjs.map(conj -> producer(conj, filter, context)).toList(), context.producer());
     }
 
-    private ResourceIterator<Producer<ConceptMap>> producers(Conjunction conjunction) {
-        return producers(conjunction, list(), defaultContext);
+    private Producer<ConceptMap> producer(Conjunction conjunction) {
+        return producer(conjunction, list(), defaultContext);
     }
 
-    private ResourceIterator<Producer<ConceptMap>> producers(Conjunction conjunction,
-                                                             List<Identifier.Variable.Name> filter,
-                                                             Context.Query context) {
-        if (!isInfer(context)) LOG.warn("Reasoning is disabled in write transactions");
+    private Producer<ConceptMap> producer(Conjunction conjunction, List<Identifier.Variable.Name> filter,
+                                          Context.Query context) {
+        if (!isInfer(context)) LOG.info("Reasoning will not run");
 
-        List<Producer<ConceptMap>> producers = new ArrayList<>();
+        Producer<ConceptMap> producer;
         Conjunction conj = logicMgr.typeResolver().resolve(conjunction);
         if (conj.isSatisfiable()) {
-            Producer<VertexMap> producer =
-                    traversalEng.producer(conj.traversal(filter), context.producer(), PARALLELISATION_FACTOR);
-            producers.add(producer.map(conceptMgr::conceptMap));
-            if (isInfer(context)) producers.add(resolve(conj));
+            if (isInfer(context)) producer = resolve(conj);
+            else {
+                producer = traversalEng.producer(conj.traversal(filter), context.producer(), PARALLELISATION_FACTOR)
+                        .map(conceptMgr::conceptMap);
+            }
         } else if (!filter.isEmpty() && iterate(filter).anyMatch(id -> conj.variable(id).isThing()) ||
                 iterate(conjunction.variables()).anyMatch(Variable::isThing)) {
             throw GraknException.of(UNSATISFIABLE_CONJUNCTION, conjunction);
         } else {
-            return single(Producers.empty());
+            return Producers.empty();
```

Comment:
The fact that this is `Producers`, plural, says again to me we need a better name 🤔